File: plugins/geo_fencing/ios/Classes/actions/InitMethodAction.swift
//Made by Vahid Rajabi

import Flutter

class InitMethodAction: MethodAction {
    private static let DATABASE_KEY = "database-path-key"
    private static let DATABASE_VERSION_KEY = "database-version-key"
    private static let TRACKER_DATABASE_NAME = "database-tracker-table"
    private static let LOCATION_DATABASE_NAME = "database-location-table"
    
    func createAction(method: FlutterMethodCall) -> ActionResult {
        guard let args = method.arguments as? [String: Any] else {
            return FailureResult(error: "Invalid arguments")
        }
        
        if let dbVersion = args[InitMethodAction.DATABASE_VERSION_KEY] as? Int {
            PluginParams.DATABASE_VERSION = dbVersion
        }
        
        initDatabase(method: method)
        
        PluginParams.LOCATION_TABLE = args[InitMethodAction.LOCATION_DATABASE_NAME] as? String
        PluginParams.TRACKER_TABLE = args[InitMethodAction.TRACKER_DATABASE_NAME] as? String
        
        return SuccessResult()
    }
    
    private func initDatabase(method: FlutterMethodCall) {
        guard let args = method.arguments as? [String: Any],
              let dbLocation = args[InitMethodAction.DATABASE_KEY] as? String,
              let dbVersion = args[InitMethodAction.DATABASE_VERSION_KEY] as? Int else {
            return
        }
        
        if !dbLocation.isEmpty {
            AppDatabase.shared.initialize(path: dbLocation, version: dbVersion)
        }
    }
    
    func hasValidArguments(method: FlutterMethodCall) -> Bool {
        guard let args = method.arguments as? [String: Any] else {
            return false
        }
        
        return args[InitMethodAction.DATABASE_KEY] != nil &&
               args[InitMethodAction.DATABASE_VERSION_KEY] != nil &&
               hasStringArgument(method: method, argument: InitMethodAction.TRACKER_DATABASE_NAME) &&
               hasStringArgument(method: method, argument: InitMethodAction.LOCATION_DATABASE_NAME)
    }
}

File: plugins/geo_fencing/android/src/main/kotlin/com/example/geo_fencing/actions/InitMethodAction.kt
package com.example.geo_fencing.actions

import android.content.Context
import com.example.geo_fencing.PluginParams
import com.example.geo_fencing.actions.result.ActionResult
import com.example.geo_fencing.actions.result.SuccessResult
import com.example.geo_fencing.database.AppDatabase
import io.flutter.plugin.common.MethodCall

class InitMethodAction(private val context: Context) : MethodAction {

    companion object {
        private const val DATABASE_KEY = "database-path-key"
        private const val DATABASE_VERSION_KEY = "database-version-key"
        private const val TRACKER_DATABASE_NAME = "database-tracker-table"
        private const val LOCATION_DATABASE_NAME = "database-location-table"
    }

    override fun createAction(method: MethodCall): ActionResult {
        PluginParams.DATABASE_VERSION = method.argument(DATABASE_VERSION_KEY)

        initDatabase(method)

        PluginParams.LOCATION_TABLE = method.argument<String>(
            LOCATION_DATABASE_NAME
        )

        PluginParams.TRACKER_TABLE = method.argument<String>(
            TRACKER_DATABASE_NAME
        )

        return SuccessResult()
    }

    private fun initDatabase(method: MethodCall) {
        val dbLocation = method.argument<String>(DATABASE_KEY)
        val dbVersion = method.argument<Int>(DATABASE_VERSION_KEY)

        if (dbLocation?.isNotEmpty() == true) {
            AppDatabase.init(context, dbLocation, dbVersion!!)
        }
    }


    override fun hasValidArguments(method: MethodCall): Boolean {
        return method.hasArgument(DATABASE_KEY)
                &&
                method.hasArgument(DATABASE_VERSION_KEY)
                &&
                hasStringArgument(method, TRACKER_DATABASE_NAME)
                &&
                hasStringArgument(method, LOCATION_DATABASE_NAME)
    }

}